File: CS/WebSite/App_Code/DataManagement.cs
using System.Data;
using System.Data.OleDb;
using System.Web;

public static class DataManagement
{
    public static DataSet RetrieveDataCache() {
        if(HttpContext.Current.Session["tasklist"] == null) {
            using(OleDbConnection connection = new OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings["DatabaseConnectionString"].ConnectionString)) {
                OleDbCommand selectCommand = new OleDbCommand("SELECT * FROM TaskList", connection);
                OleDbDataAdapter dataAdapter = new OleDbDataAdapter(selectCommand);
                DataSet dataSet = new DataSet("dataSet");

                dataAdapter.Fill(dataSet);
                dataSet.Tables[0].Constraints.Add("IDPK", dataSet.Tables[0].Columns["ID"], true);
               
                HttpContext.Current.Session["tasklist"] = dataSet;
                connection.Close();
            }
        }

        return (DataSet)HttpContext.Current.Session["tasklist"];
    }

    public static void UpdateDatabase() {
        if(HttpContext.Current.Session["tasklist"] == null)
            return;
        
        using(OleDbConnection connection = new OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings["DatabaseConnectionString"].ConnectionString)) {
            OleDbCommand selectCommand = new OleDbCommand("SELECT * FROM TaskList", connection);
            OleDbDataAdapter dataAdapter = new OleDbDataAdapter(selectCommand);
            OleDbCommandBuilder commandBuilder = PrepareCommandBuilder(dataAdapter);
            DataSet dataSet = (DataSet)HttpContext.Current.Session["tasklist"];

            dataAdapter.Update(dataSet);
            dataSet.AcceptChanges();
        }
    }

    public static void RestoreDatabase() {
        DataSet dataSet = RetrieveDataCache();

        using(OleDbConnection connection = new OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings["DatabaseCopyConnectionString"].ConnectionString)) {
            OleDbCommand selectCommand = new OleDbCommand("SELECT * FROM TaskList", connection);
            OleDbDataAdapter dataAdapter = new OleDbDataAdapter(selectCommand);
            OleDbCommandBuilder commandBuilder = PrepareCommandBuilder(dataAdapter);

            dataAdapter.AcceptChangesDuringFill = false;
            //dataAdapter.FillLoadOption = LoadOption.OverwriteChanges;
            dataAdapter.Fill(dataSet);
            //DataRowState rowState = dataSet.Tables[0].Rows[1].RowState;

            connection.ConnectionString = System.Web.Configuration.WebConfigurationManager.ConnectionStrings["DatabaseConnectionString"].ConnectionString;

            dataAdapter.Update(dataSet);
            dataSet.AcceptChanges(); 
        }
    }

    private static OleDbCommandBuilder PrepareCommandBuilder(OleDbDataAdapter dataAdapter) {
        OleDbCommandBuilder commandBuilder = new OleDbCommandBuilder(dataAdapter);

        commandBuilder.QuotePrefix = "[";
        commandBuilder.QuoteSuffix = "]";
        commandBuilder.RefreshSchema();

        return commandBuilder;
    }

}

File: VB/WebSite/App_Code/DataManagement.vb
Imports Microsoft.VisualBasic
Imports System.Data
Imports System.Data.OleDb
Imports System.Web

Public NotInheritable Class DataManagement
	Private Sub New()
	End Sub
	Public Shared Function RetrieveDataCache() As DataSet
		If HttpContext.Current.Session("tasklist") Is Nothing Then
			Using connection As New OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings("DatabaseConnectionString").ConnectionString)
				Dim selectCommand As New OleDbCommand("SELECT * FROM TaskList", connection)
				Dim dataAdapter As New OleDbDataAdapter(selectCommand)
				Dim dataSet As New DataSet("dataSet")

				dataAdapter.Fill(dataSet)
				dataSet.Tables(0).Constraints.Add("IDPK", dataSet.Tables(0).Columns("ID"), True)

				HttpContext.Current.Session("tasklist") = dataSet
				connection.Close()
			End Using
		End If

		Return CType(HttpContext.Current.Session("tasklist"), DataSet)
	End Function

	Public Shared Sub UpdateDatabase()
		If HttpContext.Current.Session("tasklist") Is Nothing Then
			Return
		End If

		Using connection As New OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings("DatabaseConnectionString").ConnectionString)
			Dim selectCommand As New OleDbCommand("SELECT * FROM TaskList", connection)
			Dim dataAdapter As New OleDbDataAdapter(selectCommand)
			Dim commandBuilder As OleDbCommandBuilder = PrepareCommandBuilder(dataAdapter)
			Dim dataSet As DataSet = CType(HttpContext.Current.Session("tasklist"), DataSet)

			dataAdapter.Update(dataSet)
			dataSet.AcceptChanges()
		End Using
	End Sub

	Public Shared Sub RestoreDatabase()
		Dim dataSet As DataSet = RetrieveDataCache()

		Using connection As New OleDbConnection(System.Web.Configuration.WebConfigurationManager.ConnectionStrings("DatabaseCopyConnectionString").ConnectionString)
			Dim selectCommand As New OleDbCommand("SELECT * FROM TaskList", connection)
			Dim dataAdapter As New OleDbDataAdapter(selectCommand)
			Dim commandBuilder As OleDbCommandBuilder = PrepareCommandBuilder(dataAdapter)

			dataAdapter.AcceptChangesDuringFill = False
			'dataAdapter.FillLoadOption = LoadOption.OverwriteChanges;
			dataAdapter.Fill(dataSet)
			'DataRowState rowState = dataSet.Tables[0].Rows[1].RowState;

			connection.ConnectionString = System.Web.Configuration.WebConfigurationManager.ConnectionStrings("DatabaseConnectionString").ConnectionString

			dataAdapter.Update(dataSet)
			dataSet.AcceptChanges()
		End Using
	End Sub

	Private Shared Function PrepareCommandBuilder(ByVal dataAdapter As OleDbDataAdapter) As OleDbCommandBuilder
		Dim commandBuilder As New OleDbCommandBuilder(dataAdapter)

		commandBuilder.QuotePrefix = "["
		commandBuilder.QuoteSuffix = "]"
		commandBuilder.RefreshSchema()

		Return commandBuilder
	End Function

End Class
